data_diff/databases/base.py

```diff
@@ -471,7 +489,10 @@ def render_select(self, parent_c: Compiler, elem: Select) -> str:
         columns = ", ".join(map(compile_fn, elem.columns)) if elem.columns else "*"
         distinct = "DISTINCT " if elem.distinct else ""
         optimizer_hints = self.optimizer_hints(elem.optimizer_hints) if elem.optimizer_hints else ""
-        select = f"SELECT {optimizer_hints}{distinct}{columns}"
+        if elem.limit_expr is not None and self.USE_TOP_INSTEAD_LIMIT:
+            select = f"SELECT TOP {elem.limit_expr} {optimizer_hints}{distinct}{columns}"
+        else:
+            select = f"SELECT {optimizer_hints}{distinct}{columns}"
```

Comment:
Having this part of SQL rendering done via `if` instead of the usual AST-based & `def render_…()` approach looks unreliable.

Can I propose a little refactoring/redesign to make (maybe as a separate PR)?

Having a fully rendering SELECT SQL in the end of this routine, send it to some new function called e.g. `limit_select(sql: str)` (a better name should be picked), which will **wrap** the provided SQL into the following one by default:

```sql
SELECT * FROM ({sql}) LIMIT {n}
```

And overrides it in the SELECT-TOP-enabled databases to be:

```sql
SELECT TOP {n} FROM ({sql})
```

We also had a few cases with unusual LIMIT syntax, which is nicely solved by this approach:

MS SQL:

```sql
SELECT * FROM ({sql}) OFFSET 0 ROWS FETCH NEXT {n} ROWS ONLY
```

Oracle:

```sql
SELECT * FROM ({sql}) FETCH NEXT {n} ROWS ONLY
```

I.e., limit the selection from a subquery instead of modifying the query itself. All databases, to my best knowledge, can optimize such cases and do not fetch the full table before limiting it.

Then remove `def offset_limit(…)` as unused & unneeded.